folding-schemes/src/folding/circuits/decider/off_chain.rs

```diff
@@ -0,0 +1,316 @@
+/// This file implements a generic offchain decider circuit.
+/// For ethereum use cases, use the `GenericOnchainDeciderCircuit`.
+/// More details can be found at the documentation page:
+/// https://privacy-scaling-explorations.github.io/sonobe-docs/design/nova-decider-offchain.html
+use ark_crypto_primitives::sponge::{
+    constraints::{AbsorbGadget, CryptographicSpongeVar},
+    poseidon::{constraints::PoseidonSpongeVar, PoseidonConfig},
+    Absorb,
+};
+use ark_ec::CurveGroup;
+use ark_r1cs_std::{
+    alloc::AllocVar, eq::EqGadget, fields::fp::FpVar, prelude::CurveVar, ToConstraintFieldGadget,
+};
+use ark_relations::r1cs::{ConstraintSynthesizer, ConstraintSystemRef, SynthesisError};
+use ark_std::{marker::PhantomData, Zero};
+
+use crate::{
+    arith::{
+        r1cs::{circuits::R1CSMatricesVar, R1CS},
+        Arith, ArithGadget,
+    },
+    folding::{
+        circuits::{
+            cyclefold::{
+                CycleFoldCommittedInstance, CycleFoldCommittedInstanceVar, CycleFoldWitness,
+            },
+            decider::{EvalGadget, KZGChallengesGadget},
+            nonnative::affine::NonNativeAffineVar,
+            CF1, CF2,
+        },
+        nova::{circuits::CommittedInstanceVar, decider_eth_circuit::WitnessVar},
+        traits::{CommittedInstanceOps, CommittedInstanceVarOps, Dummy, WitnessOps, WitnessVarOps},
+    },
+};
+
+use super::DeciderEnabledNIFS;
+
+/// Circuit that implements part of the in-circuit checks needed for the offchain verification over
+/// the Curve2's BaseField (=Curve1's ScalarField).
+pub struct GenericOffchainDeciderCircuit1<
+    C1: CurveGroup,
+    C2: CurveGroup,
+    GC2: CurveVar<C2, CF2<C2>> + ToConstraintFieldGadget<CF2<C2>>,
+    RU: CommittedInstanceOps<C1>,       // Running instance
+    IU: CommittedInstanceOps<C1>,       // Incoming instance
+    W: WitnessOps<CF1<C1>>,             // Witness
+    A: Arith<W, RU>,                    // Constraint system
+    AVar: ArithGadget<W::Var, RU::Var>, // In-circuit representation of `A`
+    D: DeciderEnabledNIFS<C1, RU, IU, W, A>,
+> {
+    pub _gc2: PhantomData<GC2>,
+    pub _avar: PhantomData<AVar>,
+    /// Constraint system of the Augmented Function circuit
+    pub arith: A,
+    pub poseidon_config: PoseidonConfig<CF1<C1>>,
+    /// public params hash
+    pub pp_hash: CF1<C1>,
+    pub i: CF1<C1>,
+    /// initial state
+    pub z_0: Vec<CF1<C1>>,
+    /// current i-th state
+    pub z_i: Vec<CF1<C1>>,
+    /// Folding scheme instances
+    pub U_i: RU,
+    pub W_i: W,
+    pub u_i: IU,
+    pub w_i: W,
+    pub U_i1: RU,
+    pub W_i1: W,
+
+    /// Helper for folding verification
+    pub proof: D::Proof,
+    pub randomness: D::Randomness,
+
+    /// CycleFold running instance
+    pub cf_U_i: CycleFoldCommittedInstance<C2>,
```

Comment:
This has been like this since day 0, always Nova+CycleFold in Sonobe (in all combinations (onchain&offchain deciders, also without deciders), also for HyperNova & ProtoGalaxy, both with CycleFold.

Reply:
I mean, is not that this specific new line forces to use CycleFold instead of 'vanilla Nova', but that the entire project is around CycleFold.

Reply:
This has been like this since day 0, always Nova+CycleFold in Sonobe (in all combinations (onchain&offchain deciders, also without deciders)), also for HyperNova & ProtoGalaxy, both with CycleFold.

I mean, is not that this specific new line forces to use CycleFold instead of 'vanilla Nova', but that the entire project is around CycleFold.
Not sure if adding a comment in this line would make it clearer, it appears in the docs & readme, and in many comments along the code.

Reply:
Do you think it is a good idea to rename the circuit to something like `CycleFoldBasedOnchainDeciderCircuit`? The point is, when someone propose a new paradigm for converting folding schemes to IVC, we can implement the corresponding decider circuit for all folding schemes, and the user can choose the best decider circuit for their use case.

Reply:
Maybe once that new paradigm appears and it is implemented we can then rename the current method if needed, but in general I try to not abstract in advance, like for the moment this struct is the only possibility in the code, and it's mentioned in the comments that it is cyclefold-based. On the current code we would be only making the name of the struct longer while not distinguishing from another struct since it would not exist yet.
If at some point we add another alternative to the OnchainDeciderCircuit that is not based on the CycleFold approach, then agree that renaming it into this name is what would make sense.

How do you see it? maybe I'm missing some part of the point, and at the end making the name longer would not be a problem neither, so if you think it makes more sense I'm happy with it too, but with the info at hand seems that it would not make things clearer or help to distinguish from other cases.

Reply:
Agreed! I am neither a big fan of premature abstraction, so let's keep the name as is until a new scheme is proposed :)